From 131b51ed99c69bfdb6680e9eec57397b8f7ff4e6 Mon Sep 17 00:00:00 2001
From: Volodymyr Bendiuga <volodymyr.bendiuga@westermo.se>
Date: Mon, 23 May 2016 11:15:06 +0200
Subject: [PATCH 21/38] route/neigh: add rtnl_neigh_get_updated() function
Organization: Westermo Network Technologies AB

Returned value can be used in calculation of expiration time for
particular fdb entry. The time returned is in ticks.

Signed-off-by: Volodymyr Bendiuga <volodymyr.bendiuga@westermo.se>
Signed-off-by: Joachim Nilsson <troglobit@gmail.com>
---
 include/netlink/route/neighbour.h | 1 +
 lib/route/neigh.c                 | 8 ++++++++
 libnl-route-3.sym                 | 1 +
 3 files changed, 10 insertions(+)

diff --git a/include/netlink/route/neighbour.h b/include/netlink/route/neighbour.h
index 609684f..0d883b3 100644
--- a/include/netlink/route/neighbour.h
+++ b/include/netlink/route/neighbour.h
@@ -55,6 +55,7 @@ extern int			rtnl_neigh_get_state(struct rtnl_neigh *);
 extern void			rtnl_neigh_unset_state(struct rtnl_neigh *,
 						       int);
 extern int                      rtnl_neigh_get_entry_state(struct rtnl_neigh *neigh);
+extern int                      rtnl_neigh_get_updated(struct rtnl_neigh *neigh);
 
 extern void			rtnl_neigh_set_flags(struct rtnl_neigh *,
 						     unsigned int);
diff --git a/lib/route/neigh.c b/lib/route/neigh.c
index 16b8bb8..9774f83 100644
--- a/lib/route/neigh.c
+++ b/lib/route/neigh.c
@@ -948,6 +948,14 @@ int rtnl_neigh_get_entry_state(struct rtnl_neigh *neigh)
 		return -1;
 }
 
+int rtnl_neigh_get_updated(struct rtnl_neigh *neigh)
+{
+	if (neigh->ce_mask & NEIGH_ATTR_CACHEINFO)
+		return neigh->n_cacheinfo.nci_updated;
+	else
+		return -1;
+}
+
 void rtnl_neigh_unset_state(struct rtnl_neigh *neigh, int state)
 {
 	neigh->n_state_mask |= state;
diff --git a/libnl-route-3.sym b/libnl-route-3.sym
index 0e9b05d..bec3f80 100644
--- a/libnl-route-3.sym
+++ b/libnl-route-3.sym
@@ -1245,4 +1245,5 @@ global:
 	rtnl_nat_set_action;
 	rtnl_nat_get_action;
 	rtnl_neigh_get_entry_state;
+	rtnl_neigh_get_updated;
 } libnl_3_4;
-- 
2.25.1

